### PR TITLE
[stable/prometheus-operator] PodEndpoint support

### PR DESCRIPTION
**Pull request by elliotfehr**
#### What this PR does / why we need it:
Starting with prometheus-operator 0.31.0 a PodMonitor resource was added to allow monitoring of pods directly without the necessity to go through Endpoints of a Service. This adds support for PodMonitors to the prometheus operator.

#### Checklist
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

### CODE REVIEW

File: stable/prometheus-operator/ci/test-values.yaml [hunk 1]
```diff
@@ -1658,3 +1678,60 @@ prometheus:
...
+    # additionalLabels: {}
+
+    ## Pod label for use in assembling a job name of the form <label value>-<port>
+    ## If no label is specified, the pod endpoint name is used.
+    ##
+    # jobLabel: ""
+
+    ## labels to transfer from the kubernetes pods to the target
+    ##
+    # podTargetLabels: ""
```
Comment (vsliouniaev) [inline stable/prometheus-operator/ci/test-values.yaml]: This doesn't appear to be configurable

File: stable/prometheus-operator/ci/test-values.yaml [hunk 2]
```diff
@@ -1658,3 +1678,60 @@ prometheus:
...
+      ##
+      # any: false
+
+      ## Explicit list of namespace names to select
+      ##
+      # matchNames: []
+
+    ## Endpoints of the selected pods to be monitored
+    ##
+    # podMetricsEndpoints: []
```
Comment (vsliouniaev) [inline stable/prometheus-operator/ci/test-values.yaml:1734]: This is called `endpoints` in the template. I think it should be renamed there to match what appears in the api spec and what you have here in values.yaml

File: stable/prometheus-operator/templates/prometheus/prometheus.yaml [hunk 3]
```diff
@@ -95,6 +95,21 @@ spec:
 {{ else }}
   serviceMonitorNamespaceSelector: {}
 {{- end }}
+{{- if .Values.prometheus.prometheusSpec.podMonitorSelector }}
+  podMonitorSelector:
```
Comment (vsliouniaev) [inline stable/prometheus-operator/templates/prometheus/prometheus.yaml:105]: I think you're missing
```
{{ toYaml .Values.prometheus.prometheusSpec.podMonitorSelector | indent 4 }}
```

File: stable/prometheus-operator/ci/test-values.yaml [hunk 4]
```diff
@@ -1658,3 +1678,60 @@ prometheus:
...
+      ##
+      #   interval: 30s
+
+      ## HTTP path to scrape for metrics
+      ##
+      #   path: /metrics
+
+      ## HTTP scheme to use for scraping
+      ##
+      #   scheme: http
```
Comment (vsliouniaev) [inline stable/prometheus-operator/ci/test-values.yaml:1753]: Might be useful to include a link to the api documentation for this section: https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#podmetricsendpoint

**Review by vsliouniaev — CHANGES_REQUESTED** [review]
Thanks a lot for the PR, there are a few typos that need to get fixed.
Could you please go over it one more time in case I've missed something?

File: stable/prometheus-operator/templates/prometheus/prometheus.yaml (hunk 3 above)
Comment (elliotfehr) [reply stable/prometheus-operator/templates/prometheus/prometheus.yaml:105]: Fixed!

File: stable/prometheus-operator/ci/test-values.yaml (hunk 4 above)
Comment (elliotfehr) [reply stable/prometheus-operator/ci/test-values.yaml:1753]: Added a link above

File: stable/prometheus-operator/templates/prometheus/podmonitors.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+      name: {{ .name }}
+      labels:
+        app: {{ template "prometheus-operator.name" $ }}-prometheus
+{{ include "prometheus-operator.labels" $ | indent 8 }}
+        {{- if .additionalLabels }}
+{{ toYaml .additionalLabels | indent 8 }}
+        {{- end }}
+    spec:
+      podMetricsEndpoints:
+{{ toYaml .endpoints | indent 8 }}
```
Comment (vsliouniaev) [inline stable/prometheus-operator/templates/prometheus/podmonitors.yaml]: This `.endpoints` should be renamed to `. podMetricsEndpoints` - same as in the spec and what appears in the values file.

File: stable/prometheus-operator/templates/prometheus/podmonitors.yaml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      jobLabel: {{ .jobLabel }}
+    {{- end }}
+    {{- if .namespaceSelector }}
+      namespaceSelector:
+{{ toYaml .namespaceSelector | indent 8 }}
+    {{- end }}
+      selector:
+{{ toYaml .selector | indent 8 }}
+    {{- if .targetLabels }}
+      targetLabels:
```
Comment (vsliouniaev) [inline stable/prometheus-operator/templates/prometheus/podmonitors.yaml]: The spec indicates this property is called `podTargetLabels`

File: stable/prometheus-operator/ci/test-values.yaml [hunk 7]
```diff
@@ -1661,3 +1685,57 @@ prometheus:
...
+      ##
+      # matchNames: []
+
+    ## Endpoints of the selected pods to be monitored
+    ## https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#podmetricsendpoint
+    ##
+    # podMetricsEndpoints: []
+      ## Name of the port this endpoint refers to
+      ## Mutually exclusive with targetPort
+      # - port: ""
```
Comment (vsliouniaev) [inline stable/prometheus-operator/ci/test-values.yaml:1737]: If you're adding documentation for some of the available properties you should add it for all available properties or just link to the API spec

**Review by vsliouniaev — CHANGES_REQUESTED** [review]

File: stable/prometheus-operator/templates/prometheus/podmonitors.yaml (hunk 6 above)
Comment (elliotfehr) [reply stable/prometheus-operator/templates/prometheus/podmonitors.yaml]: Renamed to match the operator spec.

File: stable/prometheus-operator/templates/prometheus/podmonitors.yaml (hunk 5 above)
Comment (elliotfehr) [reply stable/prometheus-operator/templates/prometheus/podmonitors.yaml]: renamed

**Comment by cgroschupp** [comment]
@vsliouniaev Can you please look over the PR again.

**Comment by elliotfehr** [comment]
@vsliouniaev Merged in master to resolve the merge conflicts in `Chart.yaml`

**Pull request merged**